EKF/Scripting: support automatic switching between GPS and optical flow

rmackay9 opened the pull request:
This PR supports automatically switching between GPS and optical flow using the EKF source switching feature and a lua script that checks the state of optical flow.

More detailed list of changes are:

- AP_Scripting gets optical flow bindings and an example script that uses it to switch EKF3 sources
- EKF3 always calculates optical flow innovations and variances so they are available to the lua script.
- other minor EKF3 changes including
    - renamed two members (varInnovOptFlow -> flowVarInnov and innovOptFlow -> flowInnov)
    - removing some unused code and a variable
    - optical flow fusion loop local variables are changed from "unsigned" to "uint8_t"
    - ofDataNew structure made local to NavEKF3_core::writeOptFlowMeas() instead of being a NavEKF3_core member
- OpticalFlow SITL driver applies the FLOW_FXSCALAR and FYSCALAR like all other drivers

One naming issue is I called the FuseOptFlow's new argument "really_fuse" perhaps this should be reversed and called "only_calc_innovations".

This has been tested in SITL but should be tested on a real vehicle before merging

IamPete1 commented on libraries/AP_Scripting/examples/ahrs-source-gps-optflow.lua:
```diff
@@ -0,0 +1,184 @@
+  end
+
+  -- check rangefinder distance threshold has been set
+  local rangefinder_thresh_dist = param:get('SCR_USER1')  -- SCR_USER1 holds rangefinder threshold
```
should use the `Parameter()` userdata for getting in the loop. You would do this in init:
```
local user1 = Parameter()
assert(user1:init('SCR_USER1'), 'get SCR_USER1 failed')
```

then call with:
`user1:get()`

IamPete1 replied:
same for all the params.

rmackay9 replied:
great, thanks.  changed to use this method for getting params.

IamPete1 commented on libraries/AP_Scripting/generator/description/bindings.desc:
```diff
@@ -263,6 +263,14 @@ singleton AP_Baro method get_pressure float
 singleton AP_Baro method get_temperature float
 singleton AP_Baro method get_external_temperature float
 
+include AP_OpticalFlow/OpticalFlow.h
```
```suggestion
include AP_OpticalFlow/OpticalFlow.h depends defined(OPTFLOW) && OPTFLOW == 1
```
I'm not %100 that define is correct on all vehicles, but there should be one that works.

rmackay9 replied:
As mentioned below, I don't think this method can work because the definition is in the vehicle layer.

IamPete1 commented on libraries/AP_Scripting/generator/description/bindings.desc:
```diff
@@ -263,6 +263,14 @@ singleton AP_Baro method get_pressure float
 
+include AP_OpticalFlow/OpticalFlow.h
+singleton OpticalFlow alias optical_flow
+singleton OpticalFlow method enabled boolean
```
```suggestion
singleton OpticalFlow  depends defined(OPTFLOW) && OPTFLOW == 1
singleton OpticalFlow method enabled boolean
```

rmackay9 replied:
I don't think this method will work because the definition is in the vehicle code instead of in the HAL_ or libraries layer.

IamPete1 replied:
maybe it will have to wait for re-work of the define, would be nice to get it working for the build server. Possibly this might work?

`!defined(OPTFLOW)  || (defined(OPTFLOW) && OPTFLOW == 1)`

rmackay9 commented:
I've updated the lua script and also removed a couple of bindings that we're not using yet.

IamPete1 commented on libraries/AP_Scripting/examples/ahrs-source-gps-optflow.lua:
```diff
@@ -0,0 +1,207 @@
+local scr_user4_param = Parameter() -- user4 param (optical flow innovation threshold)
+
+-- initialise parameters
+assert(scr_user1_param:init('SCR_USER1'), 'ahrs-source.lua: could not find SCR_USER1 parameter')
```
This prints via the lua error function so you will already get the file name and line number. 
```suggestion
assert(scr_user1_param:init('SCR_USER1'), 'could not find SCR_USER1 parameter')
```

no bid deal tho, TBH I don't think I have ever seen one of these fail so long as the param name is correct and its not like this param is vehicle dependant.

rmackay9 replied:
fixed thanks.

IamPete1 commented on libraries/AP_Scripting/examples/ahrs-source-gps-optflow.lua:
```diff
@@ -0,0 +1,207 @@
+  -- source selection from RCx_FUNCTION = 90 (EKF Source Select)
+  -- auto source from RCx_FUNCTION = 300 (Scripting1)
+  local rc_function_source = rc:find_channel_for_option(90)
+  local rc_function_auto = rc:find_channel_for_option(300)
```
you could do this in init and assert too, would lost the ability for run time config tho of course.

rmackay9 replied:
ok. I like keeping it real-time actually so that users can run the script and it basically tells them what config they've missed.

IamPete1 commented on libraries/AP_Scripting/examples/ahrs-source-gps-optflow.lua:
```diff
@@ -0,0 +1,207 @@
+  -- auto source from RCx_FUNCTION = 300 (Scripting1)
+  local rc_function_source = rc:find_channel_for_option(90)
+  local rc_function_auto = rc:find_channel_for_option(300)
+  if (rc_function_source == nil) or (rc_function_auto == nil) then
```
```suggestion
  if not rc_function_source or  not rc_function_auto then
```
just a alternate syntax, its up to you.

rmackay9 replied:
changed syntax txs!

IamPete1 commented on libraries/AP_Scripting/examples/ahrs-source-gps-optflow.lua:
```diff
@@ -0,0 +1,207 @@
+
+  -- check rangefinder distance threshold has been set
+  local rangefinder_thresh_dist = scr_user1_param:get()     -- SCR_USER1 holds rangefinder threshold
+  if (rangefinder_thresh_dist == nil) or (rangefinder_thresh_dist <= 0) then
```
I don't usually bother checking the return of the gets, the only way this will fail is if the pointer is somehow lost. If that happens it will be broken until you re-init, so you might as well just error out on trying to do the comparison on a nil.

rmackay9 replied:
ok, removed the nil check on the param get.

rmackay9 commented:
@IamPete1, thanks for the review.  I've addressed those issues and I also added an assert on the optical flow bindings to ensure that users don't try to use the script with a version of AP that doesn't include these bindings.  I accidentally used the wrong version of  AP during testing and found that it just failed silently which is no good.

priseborough commented on libraries/AP_NavEKF3/AP_NavEKF3_core.h:
```diff
@@ -1163,22 +1164,22 @@ class NavEKF3_core : public NavEKF_core_common
-    Matrix3F Tbn_flow;              // transformation matrix from body to nav axes at the middle of the optical flow sample period
-    Vector2 varInnovOptFlow;        // optical flow innovations variances (rad/sec)^2
-    Vector2 innovOptFlow;           // optical flow LOS innovations (rad/sec)
+    Vector2 flowVarInnov;           // optical flow innovations variances (rad/sec)^2
```
Nitpick, but because this is an innovation variance, I would rename to flowInnovVar

rmackay9 replied:
sure, renamed flowVarInnov to flowInnovVar.

As a side note GPS and External Nav also have their equivalent variable names backwards while Beacons has it the right way around.  Something for another day of course..

priseborough commented on libraries/AP_NavEKF3/AP_NavEKF3_OptFlowFusion.cpp:
```diff
@@ -265,12 +264,13 @@ void NavEKF3_core::EstimateTerrainOffset(const of_elements &ofDataDelayed)
+ * really_fuse should be true to actually fuse into the main filter, false to only calculate variances
 */
-void NavEKF3_core::FuseOptFlow(const of_elements &ofDataDelayed)
+void NavEKF3_core::FuseOptFlow(const of_elements &ofDataDelayed, bool really_fuse)
```
I would have it fuse by default and have it inhibit fusion by exception

..., bool inhibit_fusion)

rmackay9 replied:
ok, renamed really_fuse to inhibit_fusion and reversed the set and usage of the argument.

priseborough commented on libraries/AP_NavEKF3/AP_NavEKF3_OptFlowFusion.cpp:
```diff
@@ -693,16 +678,16 @@ void NavEKF3_core::FuseOptFlow(const of_elements &ofDataDelayed)
             // number of operations
-            for (unsigned i = 0; i<=stateIndexLim; i++) {
-                for (unsigned j = 0; j<=6; j++) {
+            for (uint8_t i = 0; i<=stateIndexLim; i++) {
```
The reason these were unsigned is that during early development of the original ekf, it gave slightly faster execution times compared to using a uint_8 for an unknown reason. I would leave these as unsigned in this PR and submit a follow up PR that changes from unsigned to uint8_t in this and all the other fusion methods once we have profiling evidence that the execution time is not adversely affected by doing so.

rmackay9 replied:
ok thanks. removed this commit to change unsigned to uint8_t

rmackay9 commented:
@priseborough, txs again for the review.  I've addressed all the issues raised, rebased on master and re-tested to confirm I didn't break anything.
